### PR TITLE
Add websocket incoming message assertion

### PR DESCRIPTION
**Pull request by huzaus**
Implemented Jmeter Assertion for websocket incoming messages which will allow to add check for any Jmeter Sampler.

### CODE REVIEW

File: src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private transient Expectation expectation;
+
+    private transient SessionsManager<String> sessionsManager = getWebsocketClient().getSessionsManager();
+
+    @Override
+    public void process() {
+        log.info("in process");
+        WebsocketSession session = sessionsManager.getOpenSession(getSessionId());
+        expectation = fromNullable(session)
+                .transform(new Function<WebsocketSession, Expectation>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java:41]: Is this java 8? If yes you should be able to replace this `Function` with lambda.

File: src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            ExpectationResult expectationResult = expectation.getResult(Long.valueOf(getTimeout()));
+
+            log.debug("Update sample result end time from " + response.getEndTime() + " to " + expectationResult.getEndTime() + ".");
+            response.setEndTime(expectationResult.getEndTime());
+
+            if (expectation instanceof WebsocketMessageRegexExpectation && sessionsManager.hasOpenSession(getSessionId())) {
+                sessionsManager
+                        .getOpenSession(getSessionId())
+                        .unregisterWebsocketIncomingMessageConsumer((WebsocketMessageRegexExpectation) expectation);
+                log.debug("Unregistered  " + expectation + " consumer.");
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java:66]: Possible approach for logging wit `{}` means: `log.debug("Unregistered {} consumer.",expectation)`

File: src/main/java/com/jmeter/websocket/plugin/elements/configurations/WebsocketSessionsManager.java [hunk 3]
```diff
@@ -19,19 +19,22 @@
...
     private static final Logger log = LoggingManager.getLoggerForClass();
-
+    private static final Supplier<WebsocketClient<String>> websocketClientSupplier = websocketClientSupplier();
     private static Supplier<CsvFileWriter> csvFileWriterSupplier;
-    private static final Supplier<WebsocketClient> websocketClientSupplier = websocketClientSupplier();
 
-    private static Supplier<WebsocketClient> websocketClientSupplier() {
-        return memoize(new Supplier<WebsocketClient>() {
+    private static Supplier<WebsocketClient<String>> websocketClientSupplier() {
+        return memoize(new Supplier<WebsocketClient<String>>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/elements/configurations/WebsocketSessionsManager.java:26]: Lambda candidate.

File: src/main/java/com/jmeter/websocket/plugin/elements/helpers/ExpectationUtils.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        return new ExpectationResultContainer(result);
+    }
+
+    public static Expectation failedExpectationWithReason(String reason) {
+        return expectationFrom(failure(reason));
+    }
+
+    public static AssertionResult assertionResultFrom(final String name, ExpectationResult expectationResult) {
+        return fromNullable(expectationResult)
+                .transform(new Function<ExpectationResult, AssertionResult>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/elements/helpers/ExpectationUtils.java:26]: Lambda candidate.

File: src/main/java/com/jmeter/websocket/plugin/elements/helpers/ExpectationUtils.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                    @Override
+                    public AssertionResult apply(ExpectationResult expectationResult) {
+                        AssertionResult assertionResult = new AssertionResult(name);
+                        assertionResult.setFailure(!expectationResult.isSuccess());
+                        assertionResult.setFailureMessage(expectationResult.getReason());
+                        return assertionResult;
+                    }
+                })
+                .or(new Supplier<AssertionResult>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/elements/helpers/ExpectationUtils.java:35]: Lambda candidate

File: src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketEndpoint.java [hunk 6]
```diff
@@ -136,15 +144,28 @@ public String toString() {
                 .toString();
     }
 
-    private Supplier<JettyWebsocket> jettyWebsocketSupplier() {
-        return new Supplier<JettyWebsocket>() {
+    private Supplier<JettySocket> jettyWebsocketSupplier(final String sessionId) {
+        return new Supplier<JettySocket>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketEndpoint.java:148]: Lambda candidate

File: src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketEndpoint.java [hunk 7]
```diff
@@ -136,15 +144,28 @@ public String toString() {
...
+                for (WebsocketMessageConsumer consumer : websocketMessageConsumers) {
+                    jettySocket.registerWebsocketIncomingMessageConsumer(consumer);
+                }
+                return jettySocket;
+            }
+        };
+    }
+
+    private Supplier<JettySessionWrapper> jettySessionWrapperSupplier(final String sessionId, final Session session, final JettySocket socket) {
+        return new Supplier<JettySessionWrapper>() {
```
Comment (RMachnik) [inline src/main/java/com/jmeter/websocket/plugin/endpoint/jetty/JettyWebsocketEndpoint.java:161]: Lambda candidate

**Review by RMachnik — APPROVED** [review]

File: src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java (hunk 1 above)
Comment (huzaus) [reply src/main/java/com/jmeter/websocket/plugin/elements/assertions/WebsocketRegexAssertion.java:41]: @rafik991 Java8 is not available.

**Pull request merged**